library/network_connections.py

```diff
@@ -428,6 +428,21 @@ def ifcfg_create(
                 connection["interface_name"],
                 " ".join(configured_coalesce),
             )
+        ethtool_ring = connection["ethtool"]["ring"]
+        configured_ring = []
+        for ring, setting in ethtool_ring.items():
+            if setting is not None:
+                if isinstance(setting, bool):
+                    setting = int(setting)
```

Comment:
this does not seem to make sense

Reply:
The data type for `ethtool.ring-rx`, `ethtool.ring-tx`, `ethtool.ring-rx-mini`, `ethtool.ring-rx-jumbo` value should be integer.
It is not necessary though, argvalidator will raise an error if the corresponding value is not integer:
```
TASK [linux-system-roles.network : Configure networking connection profiles] *****************************************************************************************************************************************************************
task path: /home/wenliang/network/tests/roles/linux-system-roles.network/tasks/main.yml:81
fatal: [Fedora-Cloud-Base-33-20210106.0.x86_64.qcow2]: FAILED! => {"_invocation": {"module_args": {"__debug_flags": "", "connections": [{"autoconnect": false, "ethtool": {"ring": {"rx": 128.6}}, "ip": {"auto6": false, "dhcp4": false}, "name": "testnic1", "type": "ethernet"}], "force_state_change": false, "ignore_errors": false, "provider": "nm"}}, "changed": false, "msg": "fatal error: configuration error: connections[0].ethtool.ring.rx: must be an integer number but is '128.6'", "stderr": "\n", "stderr_lines": [""]}
```

Reply:
Actually the above code snippet is necessary, if we define the connection profile as the following, argvalidator will not raise any error

Code:
```yaml
          vars:
            network_connections:
              - name: "{{ interface }}"
                type: ethernet
                autoconnect: no
                ip:
                  dhcp4: no
                  auto6: no
                ethtool:
                  ring:
                    rx: True
        - name: Get profile's ring options
          command: nmcli -g ethtool.ring-rx c show {{ interface }}
          register: with_ring
        - name: Assert ring options set in profile
```

Result:
```
TASK [Assert ring options set in profile] ****************************************************************************************************************************************************************************************************
task path: /home/wenliang/network/tests/playbooks/tests_ethtool_ring.yml:67
fatal: [Fedora-Cloud-Base-33-20210106.0.x86_64.qcow2]: FAILED! => {
    "assertion": "with_ring.stdout == '128'",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

Reply:
>  argvalidator will not raise any error

That is because of how `ArgValidatorNum._validate_impl()` is implemented:

```python
    def _validate_impl(self, value, name):
        v = None
        try:
            if isinstance(value, self.numeric_type):
                v = value
            else:
                v2 = self.numeric_type(value)
                # ^^^^^^^
                if isinstance(value, Util.STRING_TYPE) or v2 == value:
                    v = v2
```
It tries to convert the value into a number. If it is not desirable to convert bools here (it probably is not), then fix ArgValidatorNum's implementation (note that this is potentially a break in the API, but probably ok).


---

but I don't understand why
```python
+                if isinstance(setting, bool):
+                    setting = int(setting)
```
would ever be necessary. ArgValidatorNum should normalize this to either `None` or `int` type. If not, **that** is the bug, and needs to be investigated.

Reply:
ah, that was already in code before:

```python
                 if isinstance(setting, bool):
                     setting = int(setting)
                 configured_coalesce.append(
                     "%s %s" % (coalesce.replace("_", "-"), setting)
```
That's wrong! `ArgValidatorNum` MUST normalize the value to either `None` or `int` (and it does already, or why not?). No need to assume that `ArgValidator` would not work properly and handle cases that in fact must not happen.

Likewise,
```
+                    if setting is None:
+                        s_ethtool.option_set(nm_ring, None)
+                    else:
+                        s_ethtool.option_set_uint32(nm_ring, int(setting))
                                                            # ^^^ WRONG
```

Reply:
Rational: we want ArgValidator to normalize things and rely on it's result. It has an "API" or expected behavior, where we trust that it does perform the desired normalization. For the `ArgValidatorNum` setting, we do expect that it does what it's supposed to do, and don't try to handle cases where it would not (which would be due to a bug).

In other words, we should not add code to handle situations, that can only happen in face of a bug -- if a bug would be there, we should fix the bug instead.

Reply:
Indeed, following code in `argvalidatorNum` will normalize the value into `None` or `int` type. Even if we define the value to be `Bool`(probably will never be the case),   `self.numeric_type(Bool)` will also normalize the value to be either 1 or 0. 
Let's remove `if isinstance(setting, bool): setting = int(setting)`. Thanks for clarifying.
```
    def _validate_impl(self, value, name):
        v = None
        try:
            if isinstance(value, self.numeric_type):
                v = value
            else:
                v2 = self.numeric_type(value)
                # ^^^^^^^
                if isinstance(value, Util.STRING_TYPE) or v2 == value:
                    v = v2
```